File: src/main/java/cc/mirukuneko/realtrainmodrenewed/formation/TrainFormation.java
package cc.mirukuneko.realtrainmodrenewed.formation;

import java.util.ArrayList;
import java.util.List;

public class TrainFormation {
    private final List<String> vehicleIds;
    private String name;
    
    public TrainFormation() {
        this.vehicleIds = new ArrayList<>();
        this.name = "";
    }
    
    public TrainFormation(List<String> vehicleIds, String name) {
        this.vehicleIds = new ArrayList<>(vehicleIds);
        this.name = name != null ? name : "";
    }
    
    public void addVehicle(String vehicleId) {
        if (vehicleIds.size() < 30) {
            vehicleIds.add(vehicleId);
        }
    }
    
    public void removeVehicle(int index) {
        if (index >= 0 && index < vehicleIds.size()) {
            vehicleIds.remove(index);
        }
    }
    
    public void setVehicle(int index, String vehicleId) {
        if (index >= 0 && index < vehicleIds.size()) {
            vehicleIds.set(index, vehicleId);
        }
    }
    
    public String getVehicle(int index) {
        if (index >= 0 && index < vehicleIds.size()) {
            return vehicleIds.get(index);
        }
        return null;
    }
    
    public List<String> getAllVehicles() {
        return new ArrayList<>(vehicleIds);
    }
    
    public int getCarCount() {
        return vehicleIds.size();
    }
    
    public boolean isEmpty() {
        return vehicleIds.isEmpty();
    }
    
    public boolean isFull() {
        return vehicleIds.size() >= 30;
    }
    
    public String getName() {
        return name;
    }
    
    public void setName(String name) {
        this.name = name != null ? name : "";
    }
    
    public String getDisplayName() {
        return name.isEmpty() ? "Unnamed Formation" : name;
    }
    
    public TrainFormation copy() {
        return new TrainFormation(this.vehicleIds, this.name);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/formation/TrainFormation.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.formation

class TrainFormation @JvmOverloads constructor(
    vehicleIds: List<String> = emptyList(),
    name: String? = "",
) {
    private val vehicleIds: MutableList<String> = ArrayList(vehicleIds)
    var name: String = name ?: ""

    fun addVehicle(vehicleId: String?) {
        if (vehicleIds.size < 30 && vehicleId != null) {
            vehicleIds.add(vehicleId)
        }
    }

    fun removeVehicle(index: Int) {
        if (index in vehicleIds.indices) {
            vehicleIds.removeAt(index)
        }
    }

    fun setVehicle(index: Int, vehicleId: String) {
        if (index in vehicleIds.indices) {
            vehicleIds[index] = vehicleId
        }
    }

    fun getVehicle(index: Int): String? {
        return if (index in vehicleIds.indices) vehicleIds[index] else null
    }

    fun getAllVehicles(): List<String> = ArrayList(vehicleIds)

    fun getCarCount(): Int = vehicleIds.size

    fun isEmpty(): Boolean = vehicleIds.isEmpty()

    fun isFull(): Boolean = vehicleIds.size >= 30

    fun getDisplayName(): String = name.ifEmpty { "Unnamed Formation" }

    fun copy(): TrainFormation = TrainFormation(vehicleIds, name)
}
